### PR TITLE
Implementation of DataSegmentFinder for S3 storage

### PR DESCRIPTION
**Pull request by l15k4**
Hey,

This is my first PR to Druid, so :
1) I got inspired by the Hdfs implementation while implementing and testing it rather then writing it differently from scratch
2) I used `druid_intellij_formatting.xml`

However I didn't test it yet. I need to find a way how to use `insert-segment-to-db` with `s3` deep storage without having to upgrading druid from this branch.

### CODE REVIEW

**Comment by jaehc** [comment]
Thanks for your work. Would you check #3399? It is the same as this one.

File: extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentFinder.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
+package io.druid.storage.s3;
```
Comment (fjy) [inline extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentFinder.java]: missing license header

Comment (l15k4) [inline extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentFinder.java]: Done, however it seems to be implemented already, I missed that ...

**Comment by b-slim** [comment]
@l15k4 and @jaehc you guys need to consolidate both PRs i guess there is good amount of overlapping beteween #3399  and this PR.

**Comment by l15k4** [comment]
Hey, there is like 100% of overlapping as it is basically the same thing :-) I guess it is up to you guys. If  @jaehc's PR is tested in the field I guess it would make sense to go with his because I didn't have a chance to do so, I'll be migrating Druid no sooner than in 7-10 days.

**Comment by jaehc** [comment]
@l15k4, I have used the feature several times in my work place. It looks basically almost the same. Would you mind if #3399 goes first and then your following commits on some improvement and changes you want? Or if you give it some comments, I will apply them on it.

**Comment by l15k4** [comment]
@jaehc sure, let's go with yours, I went through it, LGTM

**Comment by nishantmonu51** [comment]
closing this as its duplicate https://github.com/druid-io/druid/pull/3399 has been merged.

**Pull request closed without merging**